add mux.HandlePath method  (v2) (#1440)
grpc-ecosystem/grpc-gateway: docs/_docs/inject_router.md

---

> category: documentation

Reviewer: ```suggestion
category: documentation
title: Adding custom routes to the mux
```

---

> # How do I use this?

Reviewer: ```suggestion
# Adding custom routes to the mux
```

> The grpc gateway v2 allows you inject custom router in grpc gateway.

Reviewer: ```suggestion
The gRPC-gateway allows you to add custom routes to the serve mux, for example if you want to support a use case that isn't supported by the grpc-gateway, like file uploads.
```

> It may become a good alternative to `gorilla/mux` or other golang router matcher.

Reviewer: Remove this please, we don't want this to be the case.

## Example

```go
package main

import (
	"context"
	"net/http"
	
	pb "github.com/grpc-ecosystem/grpc-gateway/v2/examples/internal/helloworld"
	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
)

func main() {
	ctx := context.TODO()
	mux := runtime.NewServeMux()

> 	// Register proto's GET /say/{name} url

Reviewer: ```suggestion
	// Register generated routes to mux
```

	_ = pb.RegisterGreeterHandlerServer(ctx, mux, &GreeterServer{})

> 	// Register custom router urls GET /hello/{name} url

Reviewer: ```suggestion
	// Register custom route for  GET /hello/{name}
```

	_ = mux.HandlePath("GET", "/hello/{name}", func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
		w.Write([]byte("hello " + pathParams["name"]))
	})

> 	panic(http.ListenAndServe(":8080", mux))

Reviewer: ```suggestion
	http.ListenAndServe(":8080", mux)
```

}

// GreeterServer is the server API for Greeter service.
type GreeterServer struct {
	
}

// SayHello implement to say hello
func (h *GreeterServer) SayHello(ctx context.Context, req *pb.HelloRequest) (*pb.HelloReply, error) {
	return &pb.HelloReply{
		Message: "hello " + req.Name,
	}, nil
}
```

> > Note: You may need change the `pb "github.com/grpc-ecosystem/grpc-gateway/v2/examples/internal/helloworld"` to your code

Reviewer: Remove this, this should be clear already
